### PR TITLE
Decouple training from LLM activation store / LLM evals

### PR DESCRIPTION
**Pull request by chanind**
# Description

This PR refactors the way the activations store interacts with the trainer, so the trainer is no longer coupled to the activations store or LLM evals. This should make it possible to train on any dataset / model, not just LLMs by writing a new runner. Specifically, this PR makes the following changes:

- Makes the activations store into an iterable over tensors, and changes the Trainer interface to work with any iterator over tensors as a data source.
- Removes the `layer` dim from activations from the activation store, as this was always 1 and was confusing
- Adds a separate `ActivationScaler` class to handle scaling activations, removing this from the activation store
- Improves the Huggingface wrapper so it will properly early-exit when we only want activations from a layer in the middle of the model
- Removes the `b_dec_init_method` config param. Everyone just initializes this to zero now, and that's also just a lot simpler.
- Moves evaluation into an `Evaluator` type that gets passed in to the Trainer, so our evaluations are not hardcoded to be LLM-specific
- Removes the `hook_layer` config param. This was redundant given the `hook_name`, and it's already easy to derive the `hook_layer` from the `hook_name` since TransformerLens uses standardized names for all hooks.

### Questions

I'm uncertain about the decision to move activation scaling into the Trainer. I can imagine we may have a bunch of init options in the SAE that need to init the training based on some heuristic estimated from real activations, with the scaling factor being just one of them. Does it make more sense to move this into the SAE class itself? E.g. maybe the SAE has a `setup(self, data_provider)` method (or something similar) that gets called before training starts that the SAE can use to set any hyperparams it wants using real activations, and the scaling could be calculated there? The SAE could then handle its own scaling. Maybe at the end of training, or whenever `save_model_final()` is called, the SAE class can handle folding in the scaling params to its weights or whatever.

I suppose the current implementation could be migrated to the system described above without being a breaking change for most users at least, so maybe it's fine as is?

Curious to hear others thoughts on this!

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### You have tested formatting, typing and tests

- [x] I have run `make check-ci` to check format and linting. (you can run `make format` to format code if needed.)

### CODE REVIEW

File: sae_lens/load_model.py [hunk 1]
```diff
@@ -97,6 +105,7 @@ def forward(
         loss_per_token: bool = False,
         # TODO: implement real support for stop_at_layer
```
[the remark this answers is not in the document]
Comment (chanind) [reply sae_lens/load_model.py:106]: indeed! Good catch

File: sae_lens/training/activation_scaler.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    def scale(self, acts: torch.Tensor) -> torch.Tensor:
+        if self.scaling_factor is None:
+            return acts
+        return acts * self.scaling_factor
+
+    def unscale(self, acts: torch.Tensor) -> torch.Tensor:
+        if self.scaling_factor is None:
+            return acts
+        return acts / self.scaling_factor
```
[the remark this answers is not in the document]
Comment (chanind) [reply sae_lens/training/activation_scaler.py:21]: It looks like this was correct as division?

File: sae_lens/util.py [hunk 3]
```diff
@@ -26,3 +27,14 @@ def filter_valid_dataclass_fields(
...
+
+def extract_stop_at_layer_from_tlens_hook_name(hook_name: str) -> int:
+    """Extract the stop_at layer from a HookedTransformer hook name.
+
+    Returns -1 if the hook name is not a valid HookedTransformer hook name.
+    """
+    hook_match = re.search(r"\.(\d+)\.", hook_name)
+    if hook_match is None:
+        return -1
+    return int(hook_match.group(1)) + 1
```
[the remark this answers is not in the document]
Comment (chanind) [reply sae_lens/util.py:40]: If it returns -1 it should just use the final layer (e.g. run the whole model). But agree this is a sort of hidden expectation between this function and where it's used - I'll try moving the logic closer to where it's used.

File: tests/training/test_activation_scaler.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            torch.tensor([[1.0, 0.0]]),  # norm = 1.0
+            torch.tensor([[0.0, 2.0]]),  # norm = 2.0
+        ]
+        yield from batches
+
+    mean_norm = scaler._calculate_mean_norm(
+        data_provider(), n_batches_for_norm_estimate=5
+    )
+    expected_mean = (5.0 + 10.0 + 0.0 + 1.0 + 2.0) / 5
+    assert abs(mean_norm - expected_mean) < 1e-6
```
[the remark this answers is not in the document]
Comment (chanind) [reply tests/training/test_activation_scaler.py:77]: ack good catch! This was probably written by Claude. I keep telling it not to do this but it seems to really want to for some reason...

File: sae_lens/util.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chanind) [reply sae_lens/util.py:40]: Actually it should be fine to just have this return `None`. That should be clearer I think anyway, and if `None` is returned TLens will just run the whole model which is always safe.

File: sae_lens/training/mixing_buffer.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        buffer_size: Total size of the buffer (will store buffer_size/2 activations)
+        batch_size: Size of batches to return
+        activations_loader: Iterator providing new activations
+
+    Yields:
+        Batches of activations of shape (batch_size, *activation_dims)
+    """
+
+    if buffer_size < batch_size:
+        raise ValueError("Buffer size must be greater than batch size")
```
[the remark this answers is not in the document]
Comment (chanind) [reply sae_lens/training/mixing_buffer.py]: Good call! Fixed

**Comment by chanind** [comment]
Great finds @anthonyduong9! Let me know if the current version looks alright

File: sae_lens/util.py [hunk 6]
```diff
@@ -26,3 +27,25 @@ def filter_valid_dataclass_fields(
...
+
+def extract_stop_at_layer_from_tlens_hook_name(hook_name: str) -> int | None:
+    """Extract the stop_at layer from a HookedTransformer hook name.
+
+    Returns None if the hook name is not a valid HookedTransformer hook name.
+    """
+    layer = extract_layer_from_tlens_hook_name(hook_name)
+    if layer is None:
+        return None
+    return layer + 1
```
Comment (anthonyduong9) [inline sae_lens/util.py:40]:
```suggestion
    return None if layer is None else layer + 1
```

File: sae_lens/util.py [hunk 7]
```diff
@@ -26,3 +27,25 @@ def filter_valid_dataclass_fields(
...
+
+def extract_layer_from_tlens_hook_name(hook_name: str) -> int | None:
+    """Extract the layer from a HookedTransformer hook name.
+
+    Returns None if the hook name is not a valid HookedTransformer hook name.
+    """
+    hook_match = re.search(r"\.(\d+)\.", hook_name)
+    if hook_match is None:
+        return None
+    return int(hook_match.group(1))
```
Comment (anthonyduong9) [inline sae_lens/util.py:51]:
```suggestion
    return None if hook_match is None else int(hook_match.group(1))
```

File: tests/training/test_sae_trainer.py [hunk 8]
```diff
@@ -319,30 +328,35 @@ def test_estimated_norm_scaling_factor_persistence(
...
-    ]
+    during_checkpoints = []
+    final_checkpoints = []
+    for path in checkpoint_paths:
+        if "final" not in path.parent.name:
+            with open(path) as f:
+                during_checkpoints.append(json.load(f))
+        else:
+            with open(path) as f:
+                final_checkpoints.append(json.load(f))
```
Comment (anthonyduong9) [inline tests/training/test_sae_trainer.py:358]:
```suggestion
    for path in checkpoint_paths:
        with open(path) as f:
            data = json.load(f)
        if "final" in path.parent.name:
            final_checkpoints.append(data)
        else:
            during_checkpoints.append(data)
```

We can dedupe `with open(path) as f:` and the call to `json.load()`, and define the condition in the positive, which I find easier to reason about.

**Review by anthonyduong9 — COMMENTED** [review]
We still have two references to a `hook_layer` attribute in `loading_tanh_sae.ipynb`, which I assume should be removed.

File: tests/training/test_sae_trainer.py (hunk 8 above)
Comment (chanind) [reply tests/training/test_sae_trainer.py:358]: Agreed!
